drivers/adc/adc_ite_it8xxx2.c

```diff
@@ -0,0 +1,337 @@
+
+	LOG_DBG("Channel setup succeeded!");
+
+	return 0;
+}
+
+static void adc_disable_channel(const struct device *dev, uint8_t ch)
+{
+	if (ch < CHIP_ADC_CH4) {
+		/*
+		 * for channel 0, 1, 2, and 3
+		 * bit4 ~ bit0 : indicates voltage channel[x]
+		 *               input is selected for measurement (disable)
+		 * bit 7 : W/C data valid flag
+		 */
+		*adc_ctrl_regs[ch].adc_ctrl = 0x9F;
+	} else {
+		/*
+		 * for channel 4 ~ 7.
+		 * bit4 : voltage channel disable (ch 4~7)
+		 * bit7 : W/C data valid flag
+		 */
+		*adc_ctrl_regs[ch].adc_ctrl = 0x80;
+	}
+
+	/* bit 0 : adc module disable */
+	IT83XX_ADC_ADCCFG &= ~0x01;
+
+	/* disable adc interrupt */
+	irq_disable(DT_INST_IRQN(0));
+}
+
+static int check_buffer_size(const struct adc_sequence *sequence,
+			     uint8_t active_channels)
+{
+	size_t needed_buffer_size;
+
+	needed_buffer_size = active_channels * sizeof(uint16_t);
+
+	if (sequence->options) {
+		needed_buffer_size *= (1 + sequence->options->extra_samplings);
+	}
+
+	if (sequence->buffer_size < needed_buffer_size) {
+		LOG_ERR("Provided buffer is too small (%u/%u)",
+				sequence->buffer_size, needed_buffer_size);
+		return -ENOMEM;
+	}
+
+	return 0;
+}
+
+static int adc_it8xxx2_start_read(const struct device *dev,
+			      const struct adc_sequence *sequence)
+{
+	struct adc_it8xxx2_data *data = DEV_DATA(dev);
+	int err;
+
+	data->resolution = sequence->resolution;
+	LOG_DBG("Configure resolution=%d", sequence->resolution);
+
+	if (!sequence->channels ||
+		sequence->channels & ~BIT_MASK(IT8XXX2_CHANNEL_COUNT)) {
+		LOG_ERR("Invalid selection of channels");
+		return -EINVAL;
+	}
+
+	err = check_buffer_size(sequence, 1);
+	if (err) {
+		return err;
+	}
+
+	data->buffer = sequence->buffer;
+
+	adc_context_start_read(&data->ctx, sequence);
+
+	return adc_context_wait_for_completion(&data->ctx);
+}
+
+static void adc_context_start_sampling(struct adc_context *ctx)
+{
+	struct adc_it8xxx2_data *data =
+		CONTAINER_OF(ctx, struct adc_it8xxx2_data, ctx);
+
+	data->repeat_buffer = data->buffer;
+
+	/* bit 0 : adc module enable */
+	IT83XX_ADC_ADCCFG |= 0x01;
+}
+
+static int adc_it8xxx2_read(const struct device *dev,
+			const struct adc_sequence *sequence)
+{
+	struct adc_it8xxx2_data *data = DEV_DATA(dev);
+	int err;
+
+	adc_context_lock(&data->ctx, false, NULL);
+	err = adc_it8xxx2_start_read(dev, sequence);
+	adc_context_release(&data->ctx, err);
+
+	return err;
+}
+
+static void adc_context_update_buffer_pointer(struct adc_context *ctx,
+					      bool repeat_sampling)
+{
+	struct adc_it8xxx2_data *data =
+		CONTAINER_OF(ctx, struct adc_it8xxx2_data, ctx);
+
+	if (repeat_sampling) {
+		data->buffer = data->repeat_buffer;
+	}
+}
+
+static int adc_data_valid(enum chip_adc_channel adc_ch)
+{
+	return IT83XX_ADC_ADCDVSTS & BIT(adc_ch);
+}
+
+static void adc_it8xxx2_get_sample(const struct device *dev)
+{
+	struct adc_it8xxx2_data *data = DEV_DATA(dev);
+	uint8_t valid = 0;
+
+	if (adc_data_valid(data->adc_ch)) {
```

Comment:
`data->adc_ch` contains the identifier of the last configured channel, not necessarily successfully BTW, see:
https://github.com/zephyrproject-rtos/zephyr/blob/2c545a9651c190a04394553f646a524099be6d9e/drivers/adc/adc_ite_it8xxx2.c#L68-L73
and the samples should be taken for the channels indicated in `sequence->channels` when the sampling was requested.

Is `tests/drivers/adc/adc_api` passing for this driver?

Reply:
There was a test by adc_shell.c.
Because channel setting must be done before per reading adc, so "channel_cfg->channel_id" and "sequence->channels" are same channel?

Reply:
> There was a test by adc_shell.c.

`adc_shell.c` is not a real test, the driver should be checked with `tests/drivers/adc/adc_api` to confirm that it fulfills the API requirements.

> Because channel setting must be done before per reading adc, so "channel_cfg->channel_id" and "sequence->channels" are same channel?

First a channel must be configured with `adc_channel_setup()` and assigned an [identifier](https://github.com/zephyrproject-rtos/zephyr/blob/fa8912d9f4567668929db999849753710c4d86cc/include/drivers/adc.h#L114) specified in `channel_cfg->channel_id`, then it can be used in a sequence by [selecting this channel](https://github.com/zephyrproject-rtos/zephyr/blob/fa8912d9f4567668929db999849753710c4d86cc/include/drivers/adc.h#L283) (setting the corresponding bit) in `sequence->channels`. But several channels can be configured and only some of them or even one and not necessarily the last one configured can be selected for a given sampling sequence.

Reply:
OK, I understand and I will use tests/drivers/adc/adc_api to test, thanks.